pkg/cloudprovider/provider/openstack/helper.go

```diff
@@ -361,6 +365,15 @@ func assignFloatingIPToInstance(client *gophercloud.ProviderClient, instanceID,
 		if ip, err = createFloatingIP(client, region, port.ID, floatingIPPool); err != nil {
 			return osErrorToTerminalError(err, "failed to allocate a floating ip")
 		}
+		if machine, err = machineUpdater(machine, func(m *v1alpha1.Machine) {
+			m.Finalizers = append(m.Finalizers, floatingIPReleaseFinalizer)
+			if m.Annotations == nil {
+				m.Annotations = map[string]string{}
+			}
+			m.Annotations[floatingIPIDAnnotationKey] = ip.ID
```

Comment:
My initial approach was to discover the FloatingIP on delete but that has two drawbacks:

* If the floating IP deletion fails we can't reconcile, because the instance is already gone so we cant discover it anymore
* If someone assigns an additional Floating IP to their instance, we cant tell which one to delete

So I went with storing it in an annotation instead